sdcm/remote.py

```diff
@@ -122,13 +122,35 @@ def _make_ssh_command(user="root", port=22, opts='', hosts_file='/dev/null',  #
 
 
 class CommandRunner:
+    _params = None
+
     def __init__(self, hostname, user='root', password=''):
         self.hostname = hostname
         self.user = user
         self.password = password
         self.log = SDCMAdapter(LOGGER, extra={'prefix': str(self)})
         self.connection = self._create_connection()
 
+    def get_instance_parameters(self):
+        """
+        Return instance parameters required to rebuild instance
+        """
+        return self._params
+
+    def _save_instance_parameters(self, params: dict):
```

Comment:
why do we need this?

Reply:
Currently when any thread or process need to connect to the node we take remoter and pass it from main cycle to the Thread or Process, in case of thread it does work smoothly, except you increase GIL time. But for Process it can produce unpredictible result, because depending on 'start_method' you will folowing issues :
1. For 'fork' method, which is default, you get deepcopy of remoter instance including internal states which is not valid in the context of Process and can cause problems there.
2. For 'spawn' method, you will endup with not working code, since remoter is not pickable.

Workaround is no either make remoter instance pickable, or to save all the attributes from the remoter to be able to rebuild it when Process is initiated, which is what I did. In both cases code looks the same way, except for pickling you rename them to __getstate__ and __setstate__.

Reply:
let's do something simple:

define in all the classes CommandRunner, RemoteCmdRunner, LocalCmdRunner:
```
def get_init_arguments(self):
      return dict(<all the arguments that are needed to create a new Remoter"
```
to recreate the object:
```python
    remoter_init_args = node.remoter.get_init_arguments()

    new_remoter = RemoteCmdRunner(**remoter_init_args)
```